internal/controller/pkg/signature/validate.go

```diff
@@ -0,0 +1,197 @@
+package signature
+
+import (
+	"context"
+	"crypto"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/sigstore/cosign/v2/pkg/cosign"
+	ociremote "github.com/sigstore/cosign/v2/pkg/oci/remote"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/sigstore/sigstore/pkg/fulcioroots"
+	"github.com/sigstore/sigstore/pkg/signature"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+
+	"github.com/crossplane/crossplane/apis/pkg/v1beta1"
+)
+
+const fetchCertTimeout = 30 * time.Second
+
+// Validator validates image signatures.
+type Validator interface {
+	Validate(ctx context.Context, ref name.Reference, config *v1beta1.ImageVerification, opts ...remote.Option) error
+}
+
+// NewCosignValidator returns a new CosignValidator.
+func NewCosignValidator(c client.Reader, namespace string) (*CosignValidator, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), fetchCertTimeout)
+	defer cancel()
+
+	var err error
+	opts := &cosign.CheckOpts{}
+	opts.RootCerts, err = fulcioroots.Get()
+	if err != nil {
+		return nil, errors.Errorf("cannot fetch Fulcio roots: %w", err)
+	}
+	opts.IntermediateCerts, err = fulcioroots.GetIntermediates()
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch Fulcio intermediates: %w", err)
+	}
+	opts.CTLogPubKeys, err = cosign.GetCTLogPubs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch CTLog public keys: %w", err)
+	}
+
+	opts.RekorPubKeys, err = cosign.GetRekorPubs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch Rekor public keys: %w", err)
+	}
+
+	return &CosignValidator{
+		client:    c,
+		namespace: namespace,
+
+		checkOpts: opts,
+	}, nil
+}
+
+// CosignValidator validates image signatures using cosign.
+type CosignValidator struct {
+	client    client.Reader
+	namespace string
+
+	checkOpts *cosign.CheckOpts
+}
+
+// Validate validates the image signature.
+func (c *CosignValidator) Validate(ctx context.Context, ref name.Reference, config *v1beta1.ImageVerification, opts ...remote.Option) error {
+	if config.Provider != v1beta1.ImageVerificationProviderCosign {
+		return errors.New("unsupported image verification provider")
+	}
+
+	var errs []error
+	for _, a := range config.Cosign.Authorities {
+		c.checkOpts.ClaimVerifier = cosign.SimpleClaimVerifier
+		verify := cosign.VerifyImageSignatures
+		if len(a.Attestations) > 0 {
+			c.checkOpts.ClaimVerifier = cosign.IntotoSubjectClaimVerifier
+			verify = cosign.VerifyImageAttestations
+		}
+
+		if err := c.buildCosignCheckOpts(ctx, a, ociremote.WithRemoteOptions(opts...)); err != nil {
+			errs = append(errs, errors.Errorf("authority %q: cannot build cosign check options %v", a.Name, err))
+			continue
+		}
+
+		res, ok, err := verify(ctx, ref, c.checkOpts)
```

Comment:
I notice the `verify` function is being passed `remote.Option`. Does this indicate it's going to pull the whole image in order to validate it? Will we be double-pulling images - once to unpack in the `revision` controller, and once here to validate?

Reply:
Cosign stores signatures in the OCI registry as well, so it needs to communicate at least for that. Apart from that, it communicates with the registry to pull the manifest of the image but not the all layers. 

I could kind of validate this by running cosign verify in verbose mode and asking to ChatGPT :)

```
cosign verify xpkg.upbound.io/crossplane/e2e-provider-signed-keyless:v0.3.0 --certificate-identity=turkenh@gmail.com --certificate-oidc-issuer=https://github.com/login/oauth -d
``` 

<img width="836" alt="Screenshot 2024-10-28 at 14 35 45" src="https://github.com/user-attachments/assets/fcfe32d0-73fa-4816-8ad6-1b4af017c528">

I need some more reading/research to be 100% confident though.

Reply:
Got it. If it's just the manifest that doesn't seem so bad. I wonder if this should be subject to package pull policy though? In particular if someone sets the policy to Never I think they'd be surprised to see Crossplane still trying to talk to the registry.

Reply:
I am not sure.

Package pull policy controls pulling the package content in most cases to avoid re-downloading it as there are ways to find it cached somewhere (container runtime of node, filesystem etc.). In case of signatures, I cannot think of a way to cache it somewhere, at least didn't encounter a similar knob in the [existing](https://github.com/crossplane/crossplane/blob/main/design/one-pager-package-image-config.md#prior-art-on-image-verification) image verification APIs. I haven't seen them relying on pods image pull policy to make decisions on whether to pull signatures or not. 

I would propose leaving it for now and acting based on feedback if this really turns out to be confusing.

Reply:
Yeah this might be unavoidable, but we'll need to at least document that if you turn on validation stuff will be pulled twice.